### ISSUE TITLE
tf.Variable throws TypeError on conversion to typed numpy ndarray

### ISSUE THREAD

**Issue by nfelt** [tensorflow/tensorflow#46563]
The `__array__()` method is recognized by Numpy to allow objects to be conveniently converted to a Numpy ndarray.  This doesn't work consistently for`tf.Variable` because its definition of the method uses an incorrect signature that accepts no arguments (other than `self`): https://github.com/tensorflow/tensorflow/blob/569095ba3d5a57a95595d7db685b4bb748ca7337/tensorflow/python/ops/resource_variable_ops.py#L469

This breaks Numpy usage, which expects to be able to pass a dtype argument when converting to an explicitly typed array, as in `np.array(tf.Variable(0), dtype=np.int64)`.  (The expected signature is documented a bit sparsely but can been seen [here - scroll to "PyArray_FromArrayAttr"](https://numpy.org/devdocs/reference/c-api/array.html) or by example of how `ndarray` itself defines its `__array__` method [here](https://numpy.org/devdocs/reference/generated/numpy.ndarray.__array__.html).)

Tested against the latest nightly and latest numpy:

```
numpy==1.19.5
tf-nightly==2.5.0.dev20210119
```

#### Expected behavior:

Converting `tf.Variable` with a dtype succeeds the same as `tf.constant` with a dtype, or `tf.Variable` without a dtype: 

```
$ python -c 'import tensorflow as tf; import numpy as np; np.array(tf.constant(0), dtype=np.int32)'
$ python -c 'import tensorflow as tf; import numpy as np; np.array(tf.Variable(0))'
$
```

#### Actual behavior:

It raises a TypeError:

```
$ python -c 'import tensorflow as tf; import numpy as np; np.array(tf.Variable(0), dtype=np.int32)'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
TypeError: __array__() takes 1 positional argument but 2 were given
```

---

Note that there are various other single-argument `__array__(self)` definitions within TF that should probably also be updated:
- https://github.com/tensorflow/tensorflow/blob/5a1d7415893a2e27ee6084809820cc18d2183225/tensorflow/python/keras/engine/keras_tensor.py#L276
- https://github.com/tensorflow/tensorflow/blob/0c3d87c81ed5d8ffb868b006036b6da6a0f73a1a/tensorflow/python/framework/ops.py#L852
- https://github.com/tensorflow/tensorflow/blob/0c3d87c81ed5d8ffb868b006036b6da6a0f73a1a/tensorflow/python/framework/ops.py#L1030

On that note, I'm a bit confused why `tf.constant(0)` can be converted above because the `EagerTensorBase.__array__()` definition has the same problem as `tf.Variable`.  Indeed, this can be confirmed by trying to explicitly call it like this:

```
$ python -c 'import tensorflow as tf; import numpy as np; tf.constant(0).__array__(np.int32)'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
TypeError: __array__() takes 1 positional argument but 2 were given
```

My best guess is that there's some other path through the "convert python object to numpy array" flowchart that is being taken, e.g. perhaps it's because `EagerTensorBase` implements `__len__` while `ResourceVariable` does not and so the constant is interpreted as a sequence and gets converted via that path.  However, it'd still be best if `__array__()` was consistently defined with the `dtype` argument within TF code, even for types where that path isn't reachable in normal usage today.

**Comment by saikumarchalla** [comment]
Was able to reproduce the issue in TF 2.5. Please find the gist [here](https://colab.research.google.com/gist/saikumarchalla/365bb581f878da3c4f713223ea3c0fb2/untitled91.ipynb).Thanks!

**Comment by kaczmarj** [comment]
After testing various tensorflow versions, it looks like the buggy behavior was introduced between tensorflow version 2.3.2 (which does not have the bug) and version 2.4.0 (which does have the bug).

**Comment by mohantym** [comment]
Hi @nfelt ! Issue is getting resolved in [2.8](https://colab.sandbox.google.com/gist/mohantym/ab47d10263f9fe50bf39e44526af041b/untitled91.ipynb#scrollTo=6gUtjKfliYGz) . Can we move this issue to closed status now?

**Comment by kaczmarj** [comment]
yes now that the issue is fixed in version 2.8, imho it's fine to close this issue. thank you

**Comment by mohantym** [comment]
Ok @kaczmarj ! Closing this issue as it is resolved in 2.8 . Thanks!
